pkg/client/unversioned/remotecommand/remotecommand.go

```diff
@@ -185,60 +188,87 @@ func (e *Streamer) doStream() error {
 			errorChan <- fmt.Errorf("Error executing remote command: %s", message)
 			return
 		}
+		errorChan <- nil
 	}()
-	defer errorStream.Reset()
 
+	var wg sync.WaitGroup
+
+	// set up stdin stream
 	if e.stdin != nil {
 		headers.Set(api.StreamType, api.StreamTypeStdin)
 		remoteStdin, err := conn.CreateStream(headers)
 		if err != nil {
 			return err
 		}
-		defer remoteStdin.Reset()
-		// TODO this goroutine will never exit cleanly (the io.Copy never unblocks)
-		// because stdin is not closed until the process exits. If we try to call
-		// stdin.Close(), it returns no error but doesn't unblock the copy. It will
-		// exit when the process exits, instead.
-		go cp(api.StreamTypeStdin, remoteStdin, e.stdin)
-	}
 
-	waitCount := 0
-	completedStreams := 0
+		// copy from client's stdin to container's stdin
+		go func() {
+			// if e.stdin is noninteractive, e.g. `echo abc | kubectl exec -i <pod> -- cat`, make sure
+			// we close remoteStdin as soon as the copy from e.stdin to remoteStdin finishes. Otherwise
+			// the executed command will remain running.
+			defer remoteStdin.Close()
+			cp(api.StreamTypeStdin, remoteStdin, e.stdin)
+		}()
+
+		// read from remoteStdin until the stream is closed. this is essential to
+		// be able to exit interactive sessions cleanly and not leak goroutines or
+		// hang the client's terminal.
+		//
+		// if you have an interactive session, when you type 'exit' and hit enter,
+		// the exec'd process in the container will exit. As soon as
+		// go-dockerclient finishes reading from the container's stdout, it will
+		// call Close() on the stdin io.ReadCloser, which is remoteStdin. At this
+		// point, the cp finishes, and it's safe to call wg.Done().
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer remoteStdin.Close()
```

Comment:
Yeah, a comment like this (with links to impls that we depend on) might be more helpful:
```
go-dockerclient's current hijack implementation (link to impl) waits for all three streams (stdin/stdout/stderr) to finish copying before returning. When hijack finishes copying stdout/stderr, it calls Close() on it's side of remoteStdin, which allows this copy to complete. When that happens, we should Close() on our side of remoteStdin, to allow the copy in hijack to complete, and hijack to return.
```